File: VideoApp/VideoApp/ViewControllers/Lobby/LobbyViewController.m
//
//  Copyright (C) 2019 Twilio, Inc.
//
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.
//

#import "LobbyViewController.h"
#import "TwilioVideoAppAPI.h"
#import "LocalMediaController.h"
#import "RoomViewController.h"
#import "VariableAlphaToggleButton.h"
#import "VideoApp-Swift.h"
@import TwilioVideo;

@interface LobbyViewController () <LocalMediaControllerDelegate>
@property (nonatomic, weak) IBOutlet UILabel *loggedInUser;
@property (nonatomic, weak) IBOutlet TVIVideoView *localVideoView;
@property (nonatomic, weak) IBOutlet UIView *containerView;
@property (nonatomic, weak) IBOutlet UILabel *youLabel;
@property (nonatomic, weak) IBOutlet UIImageView *noVideoImage;

@property (nonatomic, weak) IBOutlet UITextField *roomTextField;
@property (nonatomic, weak) IBOutlet VariableAlphaToggleButton *audioToggleButton;
@property (nonatomic, weak) IBOutlet VariableAlphaToggleButton *videoToggleButton;
@property (nonatomic, weak) IBOutlet UIButton *flipCameraButton;

@property (nonatomic, weak) RoomViewController *roomViewController;

@property (nonatomic, strong) LocalMediaController *localMediaController;

@end

@implementation LobbyViewController

- (BOOL)isModalInPresentation {
    // Swiping to dismiss the LobbyViewController is not desirable. Use the logout action instead.
    return YES;
}

- (void)viewDidLoad {
    [super viewDidLoad];

    // Tweak up the UI
    self.roomTextField.attributedPlaceholder = [[NSAttributedString alloc] initWithString:@"Room"
                                                                               attributes:@{ NSForegroundColorAttributeName: [UIColor lightGrayColor] }];

    self.roomTextField.text = [SwiftToObjc roomNameFromDeepLink];
    
    [self.roomTextField addTarget:self
                           action:@selector(joinRoomButtonPressed:)
                 forControlEvents:UIControlEventEditingDidEndOnExit];

    UITapGestureRecognizer *tap = [[UITapGestureRecognizer alloc]
                                   initWithTarget:self
                                   action:@selector(dismissKeyboard)];

    [self.view addGestureRecognizer:tap];

    // Make sure the video and container views are still at the back of the stack... They're being... Difficult...
    [self.view sendSubviewToBack:self.containerView];
    [self.view sendSubviewToBack:self.localVideoView];

    self.localMediaController = [LocalMediaController new];
    [self.localMediaController addDelegate:self];
    [self.localMediaController createLocalAudioTrack];
    [self.localMediaController createLocalVideoTrack];
    
    [NSNotificationCenter.defaultCenter addObserver:self selector:@selector(handleSettingChange) name:SwiftToObjc.appSettingsStoreDidChangeNotificationName object:nil];
}

- (void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];

    [self refresh];
}

- (void)handleSettingChange {
    // Restart local video so that resolution changes take place when video codec setting is changed.
    if (self.localMediaController.localVideoTrack) {
        [self.localMediaController destroyLocalVideoTrack];
        [self.localMediaController createLocalVideoTrack];
    }
    
    [self refresh];
}

- (void)refresh {
    self.loggedInUser.text = SwiftToObjc.userDisplayName;
    self.audioToggleButton.selected = !self.localMediaController.localAudioTrack;
    self.videoToggleButton.selected = !self.localMediaController.localVideoTrack;
    [self updateVideoUI:!self.localMediaController.localVideoTrack];
}

- (void)viewWillDisappear:(BOOL)animated {
    [self.localMediaController.localVideoTrack removeRenderer:self.localVideoView];

    [super viewWillDisappear:animated];
}

- (void)viewWillLayoutSubviews {
    [super viewWillLayoutSubviews];

    UILayoutGuide *guide = self.view.safeAreaLayoutGuide;
    CGFloat height = guide.layoutFrame.origin.y + 108;
    self.containerView.frame = CGRectMake(0, 0, CGRectGetWidth(self.view.bounds), height);
}

- (void)viewSafeAreaInsetsDidChange {
    [super viewSafeAreaInsetsDidChange];
    [self.view setNeedsLayout];
}

- (void)updateVideoUI:(BOOL)isMuted {
    if (isMuted) {
        self.localVideoView.hidden = YES;
        self.youLabel.hidden = NO;
        self.noVideoImage.hidden = NO;
        [self.localMediaController.localVideoTrack removeRenderer:self.localVideoView];
    } else {
        if (![self.localMediaController.localVideoTrack.renderers containsObject:self.localVideoView]) {
            [self.localMediaController.localVideoTrack addRenderer:self.localVideoView];
            self.localVideoView.mirror = self.localMediaController.shouldMirrorLocalVideoView;
        }
    }

    self.flipCameraButton.enabled = !isMuted;
}

- (IBAction)toggleAudioPressed:(id)sender {
    if (self.localMediaController.localAudioTrack) {
        [self.localMediaController destroyLocalAudioTrack];
    } else {
        [self.localMediaController createLocalAudioTrack];
    }

    self.audioToggleButton.selected = !self.localMediaController.localAudioTrack;
}

- (IBAction)toggleVideoPressed:(id)sender {
    if (self.localMediaController.localVideoTrack) {
        [self.localMediaController destroyLocalVideoTrack];
    } else {
        [self.localMediaController createLocalVideoTrack];
    }

    self.videoToggleButton.selected = !self.localMediaController.localVideoTrack;
    [self updateVideoUI:!self.localMediaController.localVideoTrack];
}

- (IBAction)flipCameraPressed:(id)sender {
    [self.localMediaController flipCamera];
}

- (IBAction)joinRoomButtonPressed:(id)sender {
    if ([self.roomTextField.text isEqualToString:@""]) {
        [self.roomTextField becomeFirstResponder];
    } else {
        [self dismissKeyboard];
        [self performSegueWithIdentifier:@"roomSegue" sender:self];
    }
}

- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    if ([segue.identifier isEqualToString:@"roomSegue"]) {
        self.roomViewController = segue.destinationViewController;
        self.roomViewController.roomName = self.roomTextField.text;
        self.roomViewController.localMediaController = self.localMediaController;
    } else if ([segue.identifier isEqualToString:@"showSettings"]) {
        [SwiftToObjc prepareForShowSettingsSegue:segue];
    }
}

-(BOOL)dismissKeyboard {
    BOOL dismissed = YES;

    if ([self.roomTextField isFirstResponder]) {
        [self.roomTextField resignFirstResponder];
    } else {
        dismissed = NO;
    }

    return dismissed;
}

- (void)handleDeepLinkedURL:(NSURL *)deepLinkedURL {
    // We only care about the user tapping the link if we are not currently in a room.
    if (!self.roomViewController) {
        NSURLComponents *urlComponents = [NSURLComponents componentsWithURL:deepLinkedURL resolvingAgainstBaseURL:NO];

        if ([urlComponents.path hasPrefix:@"/room"]) {
            NSString *roomName = urlComponents.path.lastPathComponent;
            self.roomTextField.text = roomName;

            NSString *message = [NSString stringWithFormat:@"Would you like to join room: %@?", roomName];

            typeof(self) __weak weakSelf = self;

            [[NSOperationQueue mainQueue] addOperationWithBlock:^{
                UIAlertAction *cancel = [UIAlertAction actionWithTitle:@"Cancel" style:UIAlertActionStyleCancel handler:^(UIAlertAction *action) {}];

                UIAlertAction *join = [UIAlertAction actionWithTitle:@"Join" style:UIAlertActionStyleDefault handler:^(UIAlertAction *action) {
                    typeof(self) __strong strongSelf = weakSelf;
                    [strongSelf joinRoomButtonPressed:strongSelf];
                }];

                UIAlertController *alert = [UIAlertController alertControllerWithTitle:@"Join Room?" message:message preferredStyle:UIAlertControllerStyleAlert];
                [alert addAction:cancel];
                [alert addAction:join];

                [self presentViewController:alert animated:YES completion:nil];
            }];
        }
    }
}

#pragma mark - LocalMediaControllerDelegate
- (void)localMediaControllerStartedVideoCapture:(LocalMediaController *)localMediaController {
    self.localVideoView.mirror = self.localMediaController.shouldMirrorLocalVideoView;

    self.localVideoView.hidden = NO;
    self.youLabel.hidden = YES;
    self.noVideoImage.hidden = YES;
}

@end

File: VideoApp/VideoApp/ViewControllers/Lobby/LobbyViewController.swift
//
//  Copyright (C) 2020 Twilio, Inc.
//
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.
//

import UIKit

class LobbyViewController: UIViewController {
    @IBOutlet weak var loggedInUser: UILabel!
    @IBOutlet weak var videoView: VideoView!
    @IBOutlet weak var containerView: UIView!
    @IBOutlet weak var roomTextField: UITextField!
    @IBOutlet weak var audioToggleButton: UIButton!
    @IBOutlet weak var videoToggleButton: UIButton!
    @IBOutlet weak var flipCameraButton: UIButton!
    private let roomFactory = RoomFactory()
    private let deepLinkStore: DeepLinkStoreWriting = DeepLinkStore.shared
    private let notificationCenter = NotificationCenter.default
    private var room: Room!
    private var participant: LocalParticipant { room.localParticipant }
    private var shouldRenderVideo = true

    override func viewDidLoad() {
        super.viewDidLoad()

        resetRoom()
        configureVideoView()

        roomTextField.attributedPlaceholder = NSAttributedString(
            string: "Room",
            attributes: [.foregroundColor: UIColor.lightGray]
        )
        
        if let deepLink = deepLinkStore.consumeDeepLink() {
            switch deepLink {
            case let .room(roomName): roomTextField.text = roomName
            }
        }

        roomTextField .addTarget(self, action: #selector(joinRoomButtonPressed(_:)), for: .editingDidEndOnExit)
        
        view.addGestureRecognizer(UITapGestureRecognizer(target: self, action: #selector(dismissKeyboard)))
        
        notificationCenter.addObserver(self, selector: #selector(handleRoomUpdate(_:)), name: .roomUpdate, object: nil)
        notificationCenter.addObserver(self, selector: #selector(handleSettingChange), name: .appSettingDidChange, object: nil)
        
        refresh()
    }
    
    override func viewDidDisappear(_ animated: Bool) {
        super.viewDidDisappear(animated)
        
        shouldRenderVideo = false
        configureVideoView()
    }
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)

        shouldRenderVideo = true
        refresh()
        configureVideoView()
    }
    
    override func viewWillLayoutSubviews() {
        super.viewWillLayoutSubviews()
        
        let guide = view.safeAreaLayoutGuide
        let height = guide.layoutFrame.origin.y + 108
        containerView.frame = CGRect(x: 0, y: 0, width: view.bounds.width, height: height)
    }

    override func viewSafeAreaInsetsDidChange() {
        super.viewSafeAreaInsetsDidChange()
        
        view.setNeedsLayout()
    }
    
    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
        switch segue.identifier {
        case "roomSegue":
            let roomViewController = segue.destination as! RoomViewController
            roomViewController.application = .shared
            roomViewController.viewModel = RoomViewModelFactory().makeRoomViewModel(
                roomName: roomTextField.text ?? "",
                room: room
            )
            let storyboard = UIStoryboard(name: "Main", bundle: nil)
            let statsViewController = storyboard.instantiateViewController(withIdentifier: "statsViewController") as! StatsViewController
            statsViewController.videoAppRoom = room
            roomViewController.statsViewController = statsViewController
        case "showSettings":
            let navigationController = segue.destination as! UINavigationController
            let settingsViewController = navigationController.viewControllers.first as! SettingsViewController
            settingsViewController.viewModel = GeneralSettingsViewModel(
                appInfoStore: AppInfoStoreFactory().makeAppInfoStore(),
                appSettingsStore: AppSettingsStore.shared,
                authStore: AuthStore.shared
            )
        default:
            break
        }
    }
        
    @IBAction func toggleAudioPressed(_ sender: Any) {
        participant.isMicOn = !participant.isMicOn
        refresh()
    }
    
    @IBAction func toggleVideoPressed(_ sender: Any) {
        participant.isCameraOn = !participant.isCameraOn
        refresh()
    }
    
    @IBAction func flipCameraPressed(_ sender: Any) {
        participant.cameraPosition = participant.cameraPosition == .front ? .back : .front
    }
    
    @IBAction func joinRoomButtonPressed(_ sender: Any) {
        guard let roomName = roomTextField.text, !roomName.isEmpty else {
            roomTextField.becomeFirstResponder()
            return
        }
        
        dismissKeyboard()
        performSegue(withIdentifier: "roomSegue", sender: self)
    }
    
    private func resetRoom() {
        room = roomFactory.makeRoom()
        participant.isMicOn = true
        participant.isCameraOn = true
    }
    
    @objc private func handleSettingChange() {
        resetRoom() // Pick up settings like identity and video codec
        refresh()
    }

    @objc private func handleRoomUpdate(_ notification: Notification) {
        guard let payload = notification.payload as? Room.Update else { return }
        
        switch payload {
        case let .didUpdateParticipants(participants):
            guard participants.contains(where: { $0 === participant }) else { return }

            configureVideoView()
        default:
            break
        }
    }
    
    private func configureVideoView() {
        let config = VideoView.Config(
            videoTrack: shouldRenderVideo ? participant.cameraTrack : nil,
            shouldMirror: participant.shouldMirrorCameraVideo
        )
        videoView.configure(config: config)
    }
    
    private func refresh() {
        loggedInUser.text = participant.identity
        audioToggleButton.isSelected = !participant.isMicOn
        videoToggleButton.isSelected = !participant.isCameraOn
        flipCameraButton.isEnabled = participant.isCameraOn
    }

    @objc private func dismissKeyboard() {
        roomTextField.resignFirstResponder()
    }
}
